compiler/rustc_trait_selection/src/traits/query/normalize.rs

```diff
@@ -99,8 +102,12 @@ impl<'cx, 'tcx> TypeFolder<'tcx> for QueryNormalizer<'cx, 'tcx> {
             return ty;
         }
 
+        if let Some(ty) = self.cache.get(&ty) {
+            return ty;
+        }
+
         let ty = ty.super_fold_with(self);
-        match *ty.kind() {
+        let res = (|| match *ty.kind() {
```

Comment:
is this just for outlining?

Reply:
this is mostly so we don't accidentally `return` here. I kind of like this pattern myself but I do think that it can be somewhat counterintuitive

Reply:
When using a counterintuitive pattern, please add a comment explaining it. :)